File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter06/KeypadWithViewModel/KeypadWithViewModel/KeypadViewModel.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace KeypadWithViewModel
{
    public class KeypadViewModel : INotifyPropertyChanged
    {
        string inputString = "";
        string displayText = "";
        char[] specialChars = { '*', '#' };

        public event PropertyChangedEventHandler PropertyChanged;

        // Constructor
        public KeypadViewModel()
        {
            this.AddCharacterCommand = new DelegateCommand(ExecuteAddCharacter);
            this.DeleteCharacterCommand = 
                new DelegateCommand(ExecuteDeleteCharacter, CanExecuteDeleteCharacter);
        }

        // Public properties
        public string InputString
        {
            protected set
            {
                bool previousCanExecuteDeleteChar = this.CanExecuteDeleteCharacter(null);

                if (this.SetProperty<string>(ref inputString, value))
                {
                    this.DisplayText = FormatText(inputString);

                    if (previousCanExecuteDeleteChar != this.CanExecuteDeleteCharacter(null))
                        this.DeleteCharacterCommand.RaiseCanExecuteChanged();
                }
            }

            get { return inputString; }
        }

        public string DisplayText
        {
            protected set { this.SetProperty<string>(ref displayText, value); }
            get { return displayText; }
        }

        // ICommand implementations
        public IDelegateCommand AddCharacterCommand { protected set; get; }

        public IDelegateCommand DeleteCharacterCommand { protected set; get; }

        // Execute and CanExecute methods
        void ExecuteAddCharacter(object param)
        {
            this.InputString += param as string;
        }

        void ExecuteDeleteCharacter(object param)
        {
            this.InputString = this.InputString.Substring(0, this.InputString.Length - 1);
        }

        bool CanExecuteDeleteCharacter(object param)
        {
            return this.InputString.Length > 0;
        }

        // Private method called from InputString
        string FormatText(string str)
        {
            bool hasNonNumbers = str.IndexOfAny(specialChars) != -1;
            string formatted = str;

            if (hasNonNumbers || str.Length < 4 || str.Length > 10)
            {
            }
            else if (str.Length < 8)
            {
                formatted = String.Format("{0}-{1}", str.Substring(0, 3),
                                                     str.Substring(3));
            }
            else
            {
                formatted = String.Format("({0}) {1}-{2}", str.Substring(0, 3),
                                                           str.Substring(3, 3),
                                                           str.Substring(6));
            }
            return formatted;
        }

        protected bool SetProperty<T>(ref T storage, T value,
                                      [CallerMemberName] string propertyName = null)
        {
            if (object.Equals(storage, value))
                return false;

            storage = value;
            OnPropertyChanged(propertyName);
            return true;
        }

        protected void OnPropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter06/KeypadWithViewModel/KeypadWithViewModel/KeypadViewModel.cpp
#include "pch.h"
#include "KeypadViewModel.h"

using namespace Windows::UI::Xaml::Data;
using namespace Platform;

using namespace KeypadWithViewModel;

// Constructor
KeypadViewModel::KeypadViewModel()
{
    inputString = L"";
    displayText = L"";

    addCharacterCommand = ref new DelegateCommand(
                                        ref new ExecuteDelegate(this, &KeypadViewModel::ExecuteAddCharacter));

    deleteCharacterCommand = ref new DelegateCommand(
                                        ref new ExecuteDelegate(this, &KeypadViewModel::ExecuteDeleteCharacter), 
                                        ref new CanExecuteDelegate(this, &KeypadViewModel::CanExecuteDeleteCharacter));
}

// Public properties
String^ KeypadViewModel::InputString::get()
{
    return ref new String(inputString.c_str());
}

String^ KeypadViewModel::DisplayText::get()
{
    return ref new String(displayText.c_str());
}

// ICommand implementations
IDelegateCommand^ KeypadViewModel::AddCharacterCommand::get()
{
    return addCharacterCommand;
}

IDelegateCommand^ KeypadViewModel::DeleteCharacterCommand::get()
{
    return deleteCharacterCommand;
}

// Execute and CanExecute methods
void KeypadViewModel::ExecuteAddCharacter(Object^ param)
{
    const wchar_t * newChar = dynamic_cast<String^>(param)->Data();

    bool previousCanExecuteDeleteChar = CanExecuteDeleteCharacter(nullptr);
    this->inputString.append(newChar);
    NewInputString(previousCanExecuteDeleteChar);
}

void KeypadViewModel::ExecuteDeleteCharacter(Object^ param)
{
    bool previousCanExecuteDeleteChar = CanExecuteDeleteCharacter(nullptr);
    this->inputString = this->inputString.substr(0, this->inputString.length() - 1);
    NewInputString(previousCanExecuteDeleteChar);
}

bool KeypadViewModel::CanExecuteDeleteCharacter(Object^ param)
{
    return this->inputString.length() > 0;
}

// Private methods called when input string changes
void KeypadViewModel::NewInputString(bool previousCanExecuteDeleteChar)
{
    OnPropertyChanged("InputString");
    this->displayText = FormatText(inputString);
    OnPropertyChanged("DisplayText");

    if (previousCanExecuteDeleteChar != CanExecuteDeleteCharacter(nullptr))
        DeleteCharacterCommand->RaiseCanExecuteChanged();
}

std::wstring KeypadViewModel::FormatText(std::wstring str)
{
    std::wstring formatted = str;
    bool hasNonNumbers = str.find_first_of(L"*#") != str.npos;
    
    if (hasNonNumbers || str.length() < 4 || str.length() > 10)
    {
    }
    else if (str.length() < 8)
    {
        formatted = str.substr(0, 3)
                       .append(L"-")
                       .append(str.substr(3, str.length() - 3));
    }
    else
    {
        formatted = std::wstring(L"(").append(str.substr(0, 3))
                                      .append(L")")
                                      .append(str.substr(3, 3))
                                      .append(L"-")
                                      .append(str.substr(6, str.length() - 6));
    }
    return formatted;
}

void KeypadViewModel::OnPropertyChanged(String^ propertyName)
{
    PropertyChanged(this, ref new PropertyChangedEventArgs(propertyName));
}

